pkg/controller/migmigration/task.go

```diff
@@ -98,11 +98,16 @@ func (t *Task) Run() error {
 	if t.Phase == Started || t.Phase == WaitOnResticRestart {
 		return nil
 	}
+	// If restore is already done, don't ensure migration registry creation, since
+	// the next step is to remove it.
+	skipMigRegistry := t.Phase == RestoreCompleted || t.Phase == Completed
```

Comment:
Looking at our phase list

```
	Started                 = ""
	WaitOnResticRestart     = "WaitOnResticRestart"
	ResticRestartCompleted  = "ResticRestartCompleted"
	BackupStarted           = "BackupStarted"
	BackupCompleted         = "BackupCompleted"
	BackupFailed            = "BackupFailed"
	WaitOnBackupReplication = "WaitOnBackupReplication"
	BackupReplicated        = "BackupReplicated"
	RestoreStarted          = "RestoreStarted"
	RestoreCompleted        = "RestoreCompleted"
	RestoreFailed           = "RestoreFailed"
	Completed               = "Completed"
```

Is it intended that we would _not_ skip creation of the migration registry if we were in `RestoreFailed` state?

Reply:
Also important to note that if we ever add more phases between `RestoreCompleted` and `Completed`, we will need to keep adding them to this list of phases in the conditional otherwise a registry will get recreated late in the migration process.

```
skipMigRegistry := t.Phase == RestoreCompleted || t.Phase == Completed
```

Sub-optimal IMO, guessing this is what the k8s API convention doc was trying to help us avoid when they told us not to use phases.

_But_ since we're using phases now, what do you think about defining a phase or two for creation and deletion of the MigRegistry? There may be problems with this idea, maybe you can poke some holes in it.

Reply:
I didn't worry about the failure situation since as of right now it's never set. Since I wasn't entirely sure around the dynamics of when it would be set and if/how the Run method would be re-entered after setting it, I left it out for now. However, you're right that we will need to deal with failed cases here once failure is being set. Handling it here is secondary to figuring out under which circumstances we would delete the registry on failure. The `skipMigRegistry` bit exists only to prevent us from recreating the registry after deleting it if for some reason we exit a reconcile after we start deleting registry components but before the migration is marked done.

Reply:
Adding registry-related phases is probably a good idea, but it doesn't really change the nature of the problem -- it really just changes the `RestoreCompleted` comparison above into a `DeleteRegistryStarted` comparison. Basically we need to make sure we don't run any of the registry ensure methods once the task state has moved as far as starting registry deletion -- so for now this means `RestoreCompleted` or later (or, possibly, `RestoreFailed`, but again, that depends on whether we want to delete registry on restore failure or whether we want to allow another restore attempt. That's why I'm not touching failure cases yet), but if we add registry-related states, then `RegistryDeleteStarted` or later)

Reply:
:+1:  adding explicit registry phases.

Between restic pod management and registry management, it looks being able to do >,<,= comparison on ordered phases would be handy.

Had this on the back burner for a bit: https://github.com/fusor/mig-controller/pull/109  Thoughts?

Reply:
I'll look at that PR. Ordered phases is an improvement over the raw string comparison we have now, although things might break down a bit when the workflow ceases to be completely linear. For example where do BackupFailed and RestoreFailed fit into the comparison scheme?

Reply:
Looks like we may need some combination of a comparison (i.e. `t.Phase >= RegistryDeleteStarted`) and explicit comparison for failed cases. Depending on what we do about failed tasks and the registry, we may or may not want to avoid creating the registry in one of those failure states, which may or may not match what we're doing in the non-failure comparison.